src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -314,17 +322,35 @@ private void AssertIfNull(ImmutableArray<DiagnosticData> diagnostics)
         }
 
         [Conditional("DEBUG")]
-        private void AssertIfNull(DiagnosticData diagnostic)
+        private void AssertIfNull<T>(T obj) where T : class
         {
-            if (diagnostic == null)
+            if (obj == null)
             {
                 Contract.Requires(false, "who returns invalid data?");
             }
         }
 
+        [Conditional("DEBUG")]
+        public void AssertCleanup()
+        {
+            foreach (var idAndData in _map.Values.ToList())
+            {
+                foreach (var data in idAndData.Values.ToList())
+                {
+                    AssertIfNull(data.Workspace);
+                }
+            }
+        }
+
         private struct Data : IEquatable<Data>
         {
-            public readonly Workspace Workspace;
+            private static readonly ConditionalWeakTable<Workspace, WeakReference<Workspace>> s_cache =
+                new ConditionalWeakTable<Workspace, WeakReference<Workspace>>();
+
+            // not all workspace lives whole VS session. 
+            // make sure we not strongly hold onto them.
```

Comment:
are you asking about CCW? since workspace is the key, and weakreference is holding onto same workspace instance. that will go away when workspace goes away. 

also, we have CCW all over the places. I dont think this makes workspace lives any longer than one we already have.